### PR TITLE
Add ApplicationName Validation(Azure/azure-signalr#944)

### LINKED ISSUE #944: Failed to connect to '(Primary)https://***-dev.service.signalr.net' - http status code 400
### Describe the bug

I can see multiple repeating logs from my aspnet core application, indicating a problem connecting to azure signal-r.

Here are the logs that repeat in a cycle (most recent first)

>
Failed to connect to '(Primary)https://foo-dev.service.signalr.net', will retry after the back off period. Error detail: The server returned status code '400' when status code '101' was expected.. The server returned status code '400' when status code '101' was expected.. Id: 67e26bc1-103b-4ae0-95c3-f21293b2354c
29 Jun 2020 17:43:55.860
 Transport is stopping.
29 Jun 2020 17:43:55.803
 Starting transport. Transfer mode: Binary. Url: 'wss://foo-dev.service.signalr.net/server/?hub=43e1b6f6e3f14c31b4ddbccf011c9b0d_webnotificationhub&cid=67e26bc1-103b-4ae0-95c3-f21293b2354c'.

How do I work out why this is failing with an error 400? I can't see any further details in the log output.

My startup contains code like this which prefixes hub names:

```

 s.AddSignalR((o) =>
                {
                    o.EnableDetailedErrors = true;
                })
                .AddAzureSignalR((a) =>
                {
                    a.ApplicationName = context.Tenant.GetPartitionKey(); // segregate hubs by tenant.
                    a.ConnectionString = tenantConfig[signalRConnStringKey];
                });
```

The partition key returns a GUID formatted like so:

```
return tenant?.Guid.ToString("N");
```

It seems to work locally. I only see the above error logs in an azure app service for windows containers environment.

### Further technical details
- Your Azure SignalR SDK version
>         <PackageReference Include="Microsoft.Azure.SignalR" Version="1.*" />

- Your SignalR Client SDK version

>         <PackageReference Include="Microsoft.AspNetCore.SignalR.Client" Version="3.1.5" />

### PR DESCRIPTION
**Pull request by Y-Sindo**
### Add ApplicationName Validation
 - confirm that ApplicationName should start with alphabetic characters and only contain alpha-numeric characters or underscore.

Fixes #944

### CODE REVIEW

File: src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs [hunk 1]
```diff
@@ -2,9 +2,11 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using Microsoft.Azure.SignalR.Common;
+using System.Text.RegularExpressions;
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:5]: Move `System` usings first.

File: src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs [hunk 2]
```diff
@@ -2,9 +2,11 @@
...
 
 namespace Microsoft.Azure.SignalR
 {
+
     internal static class ServiceOptionsExtensions
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:10]: Remove extra empty lines.

File: src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs [hunk 3]
```diff
@@ -14,6 +16,14 @@ public static void Validate(this ServiceOptions options)
             {
                 throw new AzureSignalRInvalidServiceOptionsException("MaxPollIntervalInSeconds", "[1,300]");
             }
+
+
+            if (!string.IsNullOrEmpty(options.ApplicationName) && !Regex.IsMatch(options.ApplicationName, "^[a-zA-Z][a-zA-Z0-9_]*$"))
+            {
+                throw new AzureSignalRInvalidServiceOptionsException("ApplicationName", "should start with alphabetic characters and only contain alpha-numeric characters or underscore.");
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:23]: >should start with alphabetic characters and only contain alpha-numeric characters or underscore. [](start = 89, length = 96)

Base message is formatted like "Property '{propertyName}' value should be {validScope}." [https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRInvalidServiceOptionsException.cs#L13](https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRInvalidServiceOptionsException.cs#L13).

Update the sentence to make it smooth.

File: src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs [hunk 4]
```diff
@@ -14,6 +16,14 @@ public static void Validate(this ServiceOptions options)
             {
                 throw new AzureSignalRInvalidServiceOptionsException("MaxPollIntervalInSeconds", "[1,300]");
             }
+
+
+            if (!string.IsNullOrEmpty(options.ApplicationName) && !Regex.IsMatch(options.ApplicationName, "^[a-zA-Z][a-zA-Z0-9_]*$"))
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:21]: Add UT?

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:21]: Excuse me, what is the meaning of "Add UT"?

Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:21]: Short for Unit Test.

Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:21]: And you can refer to [this test case](https://github.com/Azure/azure-signalr/blob/dev/test/Microsoft.Azure.SignalR.Tests/NegotiateHandlerFacts.cs#L420) to add another fact against your new rule.

**Comment by JialinXin** [comment]
Shall we update the comment in [`ServiceOptions`](https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR/ServiceOptions.cs#L29) for ApplicationName? And is this check valid for AspNet?

File: src/Microsoft.Azure.SignalR.AspNet/OwinExtensions.cs [hunk 5]
```diff
@@ -182,6 +183,11 @@ private static void RunAzureSignalRCore(IAppBuilder builder, string applicationN
                 throw new ArgumentException("Empty application name is not allowed.", nameof(applicationName));
             }
 
+            if (!Regex.IsMatch(applicationName, "^[a-zA-Z][a-zA-Z0-9_]*$"))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/OwinExtensions.cs:186]: AspNet one has no such restriction

File: src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs [hunk 6]
```diff
@@ -14,6 +15,10 @@ public static void Validate(this ServiceOptions options)
             {
                 throw new AzureSignalRInvalidServiceOptionsException("MaxPollIntervalInSeconds", "[1,300]");
             }
+            if (!string.IsNullOrEmpty(options.ApplicationName) && !Regex.IsMatch(options.ApplicationName, "^[a-zA-Z][a-zA-Z0-9_]*$"))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:18]: use a static `Regex` object with `Compiled` option

File: src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs [hunk 7]
```diff
@@ -14,6 +15,10 @@ public static void Validate(this ServiceOptions options)
...
+            {
+                throw new AzureSignalRInvalidServiceOptionsException("ApplicationName", "prefixed with alphabetic characters and only contain alpha-numeric characters or underscore");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:20]: `Application name should be prefixed with...`

File: test/Microsoft.Azure.SignalR.AspNet.Tests/RunAzureSignalRTests.cs [hunk 8]
```diff
@@ -78,7 +78,45 @@ public void TestRunAzureSignalRWithAppNameEqualToHubNameThrows()
                 Assert.Equal("App name should not be the same as hub name.", ex.Message);
             }
         }
+        /*
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.AspNet.Tests/RunAzureSignalRTests.cs:81]: why comment out?

File: test/Microsoft.Azure.SignalR.Tests/AddAzureSignalRFacts.cs [hunk 9]
```diff
@@ -400,5 +396,44 @@ public void AddAzureSignalRCustomizeEndpointsOverridesConfigValue(string default
...
+                var config = new ConfigurationBuilder().Build();
+                serviceProvider = new ServiceCollection().AddSignalR()
+                            .AddAzureSignalR(o =>
+                            {
+                                o.ApplicationName = appName;
+                            }).Services
+                            .AddSingleton(loggerFactory)
+                            .AddSingleton<IConfiguration>(config)
+                            .BuildServiceProvider();
+                if (isValid)
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.Tests/AddAzureSignalRFacts.cs:423]: separate valid and invalid cases into 2 test cases

File: src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs (hunk 7 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:20]: > Base message is formatted like "Property '{propertyName}' value should be {validScope}." https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRInvalidServiceOptionsException.cs#L13.

File: test/Microsoft.Azure.SignalR.AspNet.Tests/RunAzureSignalRTests.cs (hunk 8 above)
Comment (Y-Sindo) [reply test/Microsoft.Azure.SignalR.AspNet.Tests/RunAzureSignalRTests.cs:81]: It's an outdated file.

File: src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs [hunk 10]
```diff
@@ -1,19 +1,25 @@
...
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Text.RegularExpressions;
 using Microsoft.Azure.SignalR.Common;
 
 namespace Microsoft.Azure.SignalR
 {
     internal static class ServiceOptionsExtensions
     {
+        private static Regex appNameRegex = new Regex("^[a-zA-Z][a-zA-Z0-9_]*$", RegexOptions.Compiled);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServiceOptionsExtensions.cs:11]: add `readonly`

`appNameRegex` => capitalize

**Pull request merged**